pixy/consumer.go

```diff
@@ -0,0 +1,562 @@
+	return gc.deleteTopicConsumerCh
+}
+
+func (gc *groupConsumer) key() string {
+	return gc.group
+}
+
+func (gc *groupConsumer) start(stoppedCh chan<- dispatchTier) {
+	spawn(&gc.wg, func() {
+		defer func() { stoppedCh <- gc }()
+		gc.registry = spawnConsumerGroupRegister(gc.group, gc.config.ClientID, gc.config, gc.kazooConn)
+		var manageWg sync.WaitGroup
+		spawn(&manageWg, gc.managePartitions)
+		gc.dispatcher.start()
+		<-gc.stoppingCh
+		gc.dispatcher.stop()
+		gc.registry.stop()
+		manageWg.Wait()
+	})
+}
+
+func (gc *groupConsumer) requests() chan<- consumeRequest {
+	return gc.dispatcher.requests()
+}
+
+func (gc *groupConsumer) stop() {
+	close(gc.stoppingCh)
+	gc.wg.Wait()
+}
+
+func (gc *groupConsumer) dispatchKey(req consumeRequest) string {
+	return req.topic
+}
+
+func (gc *groupConsumer) newDispatchTier(key string) dispatchTier {
+	tc := gc.newTopicConsumer(key)
+	return tc
+}
+
+func (gc *groupConsumer) managePartitions() {
+	cid := gc.baseCID.NewChild("managePartitions")
+	defer cid.LogScope()()
+
+	topicConsumers := make(map[string]*topicConsumer)
+	var topics []string
+	var memberSubscriptions map[string][]string
+	ok := true
+	var nilOrRetryCh <-chan time.Time
+	var nilOrRegistryTopicsCh chan<- []string
+	shouldRebalance, canRebalance := false, true
+	rebalanceResultCh := make(chan error, 1)
+rebalancingLoop:
+	for {
+		select {
+		case tc := <-gc.addTopicConsumerCh:
+			topicConsumers[tc.topic] = tc
+			topics = listTopics(topicConsumers)
+			nilOrRegistryTopicsCh = gc.registry.topics()
+			continue rebalancingLoop
+		case tc := <-gc.deleteTopicConsumerCh:
+			delete(topicConsumers, tc.topic)
+			topics = listTopics(topicConsumers)
+			nilOrRegistryTopicsCh = gc.registry.topics()
+			continue rebalancingLoop
+		case nilOrRegistryTopicsCh <- topics:
+			nilOrRegistryTopicsCh = nil
+			continue rebalancingLoop
+		case memberSubscriptions, ok = <-gc.registry.membershipChanges():
+			if !ok {
+				goto done
+			}
+			nilOrRetryCh = nil
+			shouldRebalance = true
+		case err := <-rebalanceResultCh:
+			canRebalance = true
+			if err != nil {
+				log.Errorf("<%s> rebalance failed: err=(%s)", cid, err)
+				nilOrRetryCh = time.After(gc.config.Consumer.BackOffTimeout)
+				continue
+			}
+		case <-nilOrRetryCh:
+			shouldRebalance = true
+		}
+
+		if shouldRebalance && canRebalance {
+			topicConsumerCopy := make(map[string]*topicConsumer, len(topicConsumers))
+			for topic, tc := range topicConsumers {
+				topicConsumerCopy[topic] = tc
+			}
+			go gc.rebalance(topicConsumerCopy, memberSubscriptions, rebalanceResultCh)
+			shouldRebalance, canRebalance = false, false
+		}
+	}
+done:
+	var allExclusiveConsumers []*exclusiveConsumer
+	for _, topicExclusiveConsumers := range gc.exclusiveConsumers {
+		for _, ec := range topicExclusiveConsumers {
+			allExclusiveConsumers = append(allExclusiveConsumers, ec)
+		}
+	}
+	stopExclusiveConsumers(allExclusiveConsumers)
```

Comment:
I would inline `stopExclusiveConsumers` here. It's a little bit less code, and (I think) a little more clear what it's doing. It even saves a little memory and time by not allocating and resizing the `[]*exclusiveConsumer`

```go
var wg sync.WaitGroup
for _, topicExclusiveConsumers := range gc.exclusiveConsumers {
        for _, ec := range topicExclusiveConsumers {
                spawn(&wg, ec.stop)
        }
}
wg.Wait()
```

Same with the other use of `stopExclusiveConsumers`.